src/applyMiddleware.ts

```diff
@@ -23,40 +26,17 @@ import { Reducer } from './types/reducers'
  * @template Ext Dispatch signature added by a middleware.
  * @template S The type of the state supported by a middleware.
  */
-export default function applyMiddleware(): StoreEnhancer
-export default function applyMiddleware<Ext1, S>(
-  middleware1: Middleware<Ext1, S, any>
-): StoreEnhancer<{ dispatch: Ext1 }>
-export default function applyMiddleware<Ext1, Ext2, S>(
-  middleware1: Middleware<Ext1, S, any>,
-  middleware2: Middleware<Ext2, S, any>
-): StoreEnhancer<{ dispatch: Ext1 & Ext2 }>
-export default function applyMiddleware<Ext1, Ext2, Ext3, S>(
-  middleware1: Middleware<Ext1, S, any>,
-  middleware2: Middleware<Ext2, S, any>,
-  middleware3: Middleware<Ext3, S, any>
-): StoreEnhancer<{ dispatch: Ext1 & Ext2 & Ext3 }>
-export default function applyMiddleware<Ext1, Ext2, Ext3, Ext4, S>(
-  middleware1: Middleware<Ext1, S, any>,
-  middleware2: Middleware<Ext2, S, any>,
-  middleware3: Middleware<Ext3, S, any>,
-  middleware4: Middleware<Ext4, S, any>
-): StoreEnhancer<{ dispatch: Ext1 & Ext2 & Ext3 & Ext4 }>
-export default function applyMiddleware<Ext1, Ext2, Ext3, Ext4, Ext5, S>(
-  middleware1: Middleware<Ext1, S, any>,
-  middleware2: Middleware<Ext2, S, any>,
-  middleware3: Middleware<Ext3, S, any>,
-  middleware4: Middleware<Ext4, S, any>,
-  middleware5: Middleware<Ext5, S, any>
-): StoreEnhancer<{ dispatch: Ext1 & Ext2 & Ext3 & Ext4 & Ext5 }>
-export default function applyMiddleware<Ext, S = any>(
-  ...middlewares: Middleware<any, S, any>[]
-): StoreEnhancer<{ dispatch: Ext }>
-export default function applyMiddleware(
-  ...middlewares: Middleware[]
-): StoreEnhancer<any> {
-  return (createStore: StoreCreator) => <S, A extends AnyAction>(
-    reducer: Reducer<S, A>,
+export default function applyMiddleware<
+  S = any,
+  M extends Middleware = Middleware
```

Comment:
Why does this need to be generic? In what case would our middleware not be a `Middleware`? Either we don't need this, or we need to add a test for this generic.

Reply:
If it's not generic, I can't _infer_ off of it; rather, I do get errors if I try to do `typeof middlewares extends Middleware<any, any, infer D>`.

```
Type '(createStore: StoreEnhancerStoreCreator<any, never>) => <S = any, A extends Action<any> = AnyAction>(reducer: Reducer<S, A>, ...args: any[]) => any' is not assignable to type 'StoreEnhancer<never, S>'.
  Type '<S = any, A extends Action<any> = AnyAction>(reducer: Reducer<S, A>, ...args: any[]) => any' is not assignable to type 'StoreEnhancerStoreCreator<never, S>'.
    Type 'any' is not assignable to type 'never'.ts(2322)
```

It will always be a `Middleware`, which is why I said it must `extend Middleware`, but there's no reason someone can't create an interface which extends `Middleware` and use that too.

Reply:
Also, we already have tests against this generic in the existing `applyMiddleware.spec.ts` which pass just fine for some reason; though, I am seeing red underlines in my editor (even before my changes here). I've fixed those red-line errors in other pieces of my work, but once I started breaking things apart, it's a bit harder to compose these things into smaller chunks w/o some overlap.